protocol/directory.go

```diff
@@ -0,0 +1,48 @@
+package protocol
+
+import (
+	"log"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type ConiksUserPolicies struct {
+	Username               string
+	AllowUnsignedKeychange bool
+	AllowPublicLookup      bool
+}
+
+type ConiksDirectory struct {
+	*merkletree.PAD
+}
+
+func InitDirectory(policies merkletree.Policies, signKey sign.PrivateKey,
+	dirSize uint64) *ConiksDirectory {
+	pad, err := merkletree.NewPAD(policies, signKey, dirSize)
+	if err != nil {
+		panic(err)
+	}
+	return &ConiksDirectory{pad}
+}
+
+func (d *ConiksDirectory) Register(uname string, key []byte) (
+	*merkletree.AuthenticationPath, *merkletree.TemporaryBinding, ErrorCode) {
+	// check whether the name already exists
+	// in the directory before we register
+	ap, err := d.Lookup(uname)
+	if err != nil {
+		return nil, nil, ErrorInternalServer
+	}
+	if ap.Leaf.Value() != nil {
```

Comment:
Unique **to the `uname`**, yes.  Your comment says,

> check whether the name already exists

It's not enough to check that the `Value()` is `nil`, because the `ap.Leaf` can be one of three things,

 - empty
 - another leaf with a shared prefix, but `i != j`
 - a leaf with `i == j`  // the error case

This is the same as the proof of absence / inclusion, but we don't need to verify the auth path, because we're the directory here, not a client, so we trust ourself.